File: code/back-end/microservices/alphavantage-ativos/jobs/stock.py
import pandas

from AlphaVantage.alphavantage import AlphaVantage
from databases.stock import DB
from databases.db_redis import DB_Redis
import logging
logging.basicConfig(level=logging.DEBUG, format='%(asctime)s - %(name)s - %(levelname)s - %(message)s')


def update_stock_data():
    db = DB()

    alpha_vantage = AlphaVantage()

    redis = DB_Redis()

    stocks = db.get_all_stocks()

    for stock in stocks:
        id_stock = stock[0]
        nome = stock[1]
        try:
            data: pandas.DataFrame = alpha_vantage.get_stock_data(nome)
            fist_row: pandas.DataFrame = data.iloc[0].values

            opening_price = float(fist_row['1. open'])
            closing_price = float(fist_row['4. close'])
            volume = float(fist_row['5. volume'])
            current_price = float(fist_row['3. low'])

            logging.debug(f"{opening_price} - {closing_price} - {volume} - {current_price}")

            db.update_stock(id_stock, current_price, volume, opening_price, closing_price)
            redis.save_stock_data(data, nome)
        except Exception as e:
            logging.error(f"Erro ao buscar dados do ativo {nome} com as chaves disponíveis")


    db.close()
    logging.debug("Fechando conexão com o banco de dados")

File: code/front-end/kosmos_frontend/src/app/models/stock.ts
export class Stock {

  public id: number;
  public nome: string;
  public descricao: string;
  public cotacao: number;
  public volume: number;
  public created_at: string;
  public updated_at: string;
  public opening_price: number;
  public closing_price: number;

  constructor(id: number, nome:string, descricao: string, cotacao: number, volume: number, created_at: string, updated_at: string, opening_price: number, closing_price: number) {

    this.cotacao = cotacao;
    this.created_at = created_at;
    this.descricao = descricao;
    this.id = id;
    this.nome = nome;
    this.updated_at = updated_at;
    this.volume = volume;
    this.opening_price = opening_price;
    this.closing_price = closing_price;

  }
}
